main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer.scala

```diff
@@ -216,15 +200,56 @@ object OccurrenceAnalyzer {
   /**
     * Returns true if `defn` occurs in `ctx`.
     */
-  private def isSelfRecursive(defn: OccurrenceAst.Def, ctx: ExpContext): Boolean = ctx.defs.get(defn.sym) match {
-    case None => false
-    case Some(o) => o match {
-      case Occur.Dead => false
-      case Occur.Once => true
-      case Occur.OnceInLambda => true
-      case Occur.OnceInLocalDef => true
-      case Occur.Many => true
-      case Occur.ManyBranch => true
-    }
+  private def isSelfRecursive(occur: Occur): Boolean = occur match {
+    case Occur.Dead => false
+    case Occur.Once => true
+    case Occur.OnceInLambda => true
+    case Occur.OnceInLocalDef => true
+    case Occur.Many => true
+    case Occur.ManyBranch => true
   }
+
+  /**
+    * Companion object for [[LocalContext]].
+    */
+  private object LocalContext {
+
+    /**
+      * Returns a fresh [[LocalContext]].
+      */
+    def mk(): LocalContext = new LocalContext(new AtomicInteger(0), new AtomicInteger(0))
+
+  }
+
+  /**
+    * A local context, scoped for each function definition.
+    * No requirements on thread-safety since it is scoped.
+    *
+    * @param localDefs The number of declared [[OccurrenceAst.Expr.LocalDef]]s in the expression.
+    *                  Must be mutable.
+    * @param size      The total number of subexpressions (including the expression itself).
+    *                  Must be mutable.
+    */
+  private case class LocalContext(localDefs: AtomicInteger, size: AtomicInteger)
+
+  /**
+    * Companion object for [[SharedContext]].
+    */
+  private object SharedContext {
+
+    /**
+      * Returns a fresh [[SharedContext]].
+      */
+    def mk(): SharedContext = new SharedContext(new ConcurrentLinkedQueue())
+
+  }
+
+  /**
+    * A global shared context. Must be thread-safe.
+    *
+    * @param defs A map from function symbols to occurrence information.
+    *             If the map does not contain a certain symbol, then the symbol is [[Dead]].
+    */
+  private case class SharedContext(defs: ConcurrentLinkedQueue[(DefnSym, Occur)])
```

Comment:
What values can Occur actually take on?

Also, it would seem this list can be very long... Making it less efficient.

Reply:
I suppose we can use a hash map